swarm.js

```diff
@@ -14,87 +13,69 @@ const CLOSED = Symbol('NETWORK_CLOSED')
+      socket: this._onsocket.bind(this),
+      close: this._onclose.bind(this)
+    })
 
-    this._tcp = net.createServer()
-    this._utp = utp(opts)
     this._status = NOT_BOUND
-
-    this.discovery = null
-
-    this._utp.on('connection', this._onsocket.bind(this, false))
-    this._tcp.on('connection', this._onsocket.bind(this, true))
   }
 
   address () {
-    return this._tcp.address()
+    return this.network.address()
   }
-
-  _onsocket (isTCP, s, info) {
-    if (!info) info = peerInfo(null)
-    info.connected(s, isTCP)
-    onclose(s, info.disconnected.bind(info))
-    this.emit('connection', s, info)
+  _onclose () {
+    this._status = CLOSED
+    this.emit('close')
+  }
+  _onsocket (socket, isTCP, info = peerInfo(null)) {
+    info.connected(socket, isTCP)
+    this.emit('connection', socket, info)
   }
 
-  listen (port, onlistening) {
-    if (typeof port === 'function') return this.listen(0, port)
-
+  listen (port, cb) {
     assert(this._status === NOT_BOUND, 'Network already bound')
+    if (typeof port === 'function') {
+      cb = port
+      port = 0
+    }
     this._status = BINDING
-
-    if (!port) port = 0
-    if (onlistening) this.once('listening', onlistening)
-
-    const self = this
-    bind(0)
-
-    function bind (tries) {
-      listenBoth(self._tcp, self._utp, port, function (err) {
-        if (!err) return bound()
-        listenBoth(self._tcp, self._utp, 0, function (err) {
-          if (!err) return bound()
-          onerror(err)
-        })
-      })
-
-      function onerror (err) {
-        if (self._status === CLOSING) {
-          self._status = CLOSED
-          self.emit('close')
+    const bound = (err) => {
+      if (err) {
+        if (this._status === CLOSING) {
+          this._status = CLOSED
+          this.emit('close')
           return
         }
-
-        if (++tries < 3) return bind(tries)
-        self._status = NOT_BOUND
-        self.emit('error', err)
+        this._status = NOT_BOUND
+        this.emit('error', err)
+        if (cb) cb(err)
+        return
       }
+      if (this._status === CLOSING) return
+      this._status = BOUND
+      this.emit('listening')
+      if (cb) cb()
     }
-
-    function bound () {
-      if (self._status === CLOSING) return self._destroy()
-
-      const opts = Object.assign({ socket: self._utp }, self.options)
-
-      self._status = BOUND
-      self.discovery = discovery(opts)
-      self.emit('listening')
-    }
+    this.network.bind(port, bound)
   }
 
-  destroy (onclose) {
+  destroy (cb) {
     if (this._status === CLOSED) {
-      if (onclose) process.nextTick(onclose)
+      if (cb) process.nextTick(cb)
       return
     }
 
-    if (onclose) this.once('close', onclose)
+    if (cb) this.once('close', cb)
 
-    if (this._status === BOUND) {
-      this._destroy()
+    if (this._status === BINDING || this._status === BOUND) {
+      this._status = CLOSING
+      this.network.close(cb)
```

Comment:
guts handles all of this stuff. rename destroy to close and simply call .network.close(cb)